src/refdb.c

```diff
@@ -68,6 +68,16 @@ int git_refdb_set_backend(git_refdb *db, git_refdb_backend *backend)
 {
 	GIT_ERROR_CHECK_VERSION(backend, GIT_REFDB_BACKEND_VERSION, "git_refdb_backend");
 
+	if (!backend->exists || !backend->lookup || !backend->iterator ||
+	    !backend->write || !backend->rename || !backend->del ||
+	    !backend->has_log || !backend->ensure_log || !backend->free ||
+	    !backend->reflog_read || !backend->reflog_write ||
+	    !backend->reflog_rename || !backend->reflog_delete ||
+	    (backend->lock && !backend->unlock)) {
+		git_error_set(GIT_ERROR_REFERENCE, "incomplete refdb backend implementation");
+		return GIT_EINVALID;
```

Comment:
Good point. I didn't want to fail too hard, since we were missing the check in the first place, but maybe an assert is warranted.

Reply:
But if we used an assert, then we wouldn't error at all on non-debug builds

Reply:
Yeah but this is a truism throughout our codebase.  We generally assert in places that are only a result of the consumer of libgit2 missing the preconditions for our api or writing bad or incomplete code.

Reply:
I think I prefer runtime checks/error return to static asserts for things like this, as I'd be raising exceptions (`NSInternalInconsistencyException` FTW) if I had those. But asserts don't really work that way. I just had a feeling that the user might not be in control of that code as well (think copy-pasta from libgit2-backends).

As another data point, struct version checks are currently runtime errors, not asserts.

Reply:
I think moving asserts to something else is a reasonable concern.  But I think it should be a separate PR instead of trying to figure out what that next thing is in this one, and I think we should move the entire code base en masse.

So I think that we should `assert` for now so that we're consistent with the rest of the code base and so that it's easier to mechanically replace when we decide on what we _do_ want to do.

Reply:
If that's possible, I'd like to keep it as it: as this is kinda a backward incompatible change about 1) something we forgot to enforce and 2) it's only to be more helpful, as right now "bad" users would eat a segfault anyway, I feel like it would be more helpful for unsuspecting end-users to have a normal "error" (even if it's not quite expected).

To be clear, I'm trying to [fixing](https://github.com/libgit2/libgit2/compare/master...tiennou:tests/dont-delete-reflog-on-rename) [the](https://github.com/libgit2/libgit2/compare/master...tiennou:tests/ref-rename-reflog) [few](https://github.com/libgit2/libgit2/issues/5003) [bugs](https://github.com/libgit2/libgit2/compare/master...tiennou:tests/ref-rename-failure) in the refdb layer for which I have an almost 2-years old patch set, so if it's really that contentious, I'll open the issue about the general cleanup and drop it from here — with the missing version check is in, at least this struct is safe for 1.0.